### PR TITLE
Introduced a compression level to properly save best checkpoints

### PR DESCRIPTION
**Pull request by ljaljushkin**
Compression level is considered on choosing best accuracy and saving checkpoint.

If current accuracy less then the best one, checkpoint still can be best if current compression level is bigger then best one. Compression levels in ascending: NONE, PARTIAL, FULL.

### CODE REVIEW

**Comment by vshampor** [comment]
Should document this new functionality in Usage.md (or a more suitable location), so that users are made aware of it when trying to learn basic usage of NNCF.

File: nncf/pruning/filter_pruning/algo.py [hunk 1]
```diff
@@ -153,3 +152,12 @@ def export_model(self, filename, *args, **kwargs):
         """
         self._apply_masks()
         super().export_model(filename, *args, **kwargs)
+
+    def compression_level(self) -> CompressionLevel:
+        target_pruning_level = self.scheduler.pruning_target
```
Comment (ljaljushkin) [inline nncf/pruning/filter_pruning/algo.py:157]: @mkaglins please take a look.

Comment (mkaglins) [reply nncf/pruning/filter_pruning/algo.py:157]: Looks good to me.

File: examples/classification/main.py [hunk 2]
```diff
@@ -218,9 +222,14 @@ def train(config, compression_ctrl, model, criterion, is_inception, lr_scheduler
             # evaluate on validation set
             acc1, _ = validate(val_loader, model, criterion, config)
 
-        # remember best acc@1 and save checkpoint
-        is_best = acc1 > best_acc1
+        compression_level = compression_ctrl.compression_level()
+        # remember best acc@1, considering compression level. If current acc@1 less then the best acc@1, checkpoint
+        # still can be best if current compression level is bigger then best one. Compression levels in ascending
+        # order: NONE, PARTIAL, FULL.
+        is_best_by_accuracy = acc1 > best_acc1 and compression_level == best_compression_level
```
[the remark this answers is not in the document]
Comment (ljaljushkin) [reply examples/classification/main.py:229]: No, `is_compressed()` won't work properly. 

There're 3 levels of compressity: uncompressed (U), partly (P) compressed and fully (F) compressed. We can combine them into 2 groups, compressed and uncompressed, in 2 ways: 
1) uncompressed = U or P, compressed = F 
2) uncompressed = U, compressed = P or F


First option is bad for RB sparsity. Let's say fully compressed model is the model with ratio of zero elements equals to a target sparsity rate (e.g. 61%). In that case, `is_compressed` may always return `False`, even if actual number of zeros is e.g. 60.9% and it may lead to incorrect choosing of best checkpoint just based on the value of accuracy, e.g. the best accuracy is achieved on 30% of sparsity and checkpoint with 60.9% of sparsity with lower accuracy won't be best and will be lost.
Moreover we say that partly compressed model is not actually compressed, which can be misleading for user.
`is_compressed` will return `False` for intermediate level of sparsity, pruning, for models with quantized activations. I don't think this is the right choice.

Let's consider second option, where there's no difference between partly and fully compressed models. Then again the model with intermediate level of sparsity (e.g. 30%) with the best accuracy (e.g. 71%) will be saved as the best checkpoint, even if there's model with slightly worse accuracy (e.g. 70.5%) but with final level of sparsity (e.g. 61%).
This is applicable for almost all algorithms: magnitude and RB sparsity, filter pruning, staged quantization and binarization.


And yes, logic related to checkpoint resuming can be removed. I will delete it from samples.

[the remark this answers is not in the document]
Comment (AlexKoff88) [reply examples/classification/main.py:229]: My suggestion is to define `compressed` and `noncompressed` using the number of epochs allocated for the compression period (we believe that rest of the epochs are used for fine-tuning). This is quite a simple conception. Here `compressed` means that we finished scheduling and tuning hyperparameters of the compression method.

`Partially compressed `  is not obvious to me and complicates API and user's code. Is there any other reason why we need it?

[the remark this answers is not in the document]
Comment (ljaljushkin) [reply examples/classification/main.py:229]: With your suggestions it's not obvious why model is not compressed with intermediate level of compression.
30% of sparsity is not compressed model, but with 60% compressed. It looks confusing.

Moreover, number of epoch is not an indicator that model reached target compression level - e.g. ratio of zero elements. RB sparsity may require extra fine-tuning after some number of epochs when scheduler set target sparsity. 
I.e. controller said it's compressed, but actually it's not.

Partially compressed definition allows to remove such mismatch.

[the remark this answers is not in the document]
Comment (AlexKoff88) [reply examples/classification/main.py:229]: Does it mean in your proposal that RB sparsity unlikely achieves `FULL` status? I would say that this is also not obvious for the user if he/she is going to pull the status and follow up it.

[the remark this answers is not in the document]
Comment (ljaljushkin) [reply examples/classification/main.py:225]: RB sparsity doesn't work well with all scenarios, because of specificity of regularization method - it doesn't guarantee target sparsity rate.

But all other "staged" methods (magnitude sparsity, binarization, staged quantization and filter pruning) work better with separation between uncompressed and partially compressed.

Furthermore with `is_compressed()` code would be the same, no simplifications for the user:
```
    is_compressed = compression_ctrl.is_compressed()
    is_best_by_accuracy = acc1 > best_acc1 and is_compressed == was_previously_compressed
    is_best = is_best_by_accuracy or compression_level != was_previously_compressed
    best_acc1 = max(acc1, best_acc1)
    was_previously_compressed = is_compressed
```

with compression levels:

```
    compression_level = compression_ctrl.compression_level()
    is_best_by_accuracy = acc1 > best_acc1 and compression_level == best_compression_level
    is_best = is_best_by_accuracy or compression_level > best_compression_level
    best_acc1 = max(acc1, best_acc1)
    best_compression_level = max(compression_level, best_compression_level)
```

[the remark this answers is not in the document]
Comment (AlexKoff88) [reply examples/classification/main.py:225]: With `is_compressed()` the logic of saving the best checkpoint can be simplified to the following one:
- We don't save or track best accuracy and checkpoint until `is_compressed` is true
- When it gets true we start doing that

```
if compression_ctrl.is_compressed() and acc1 > best_acc1:
    best_acc1 = acc1
    save(...)
```

[the remark this answers is not in the document]
Comment (ljaljushkin) [reply examples/classification/main.py:225]: But still the issue is not resolved for all algos except RB Sparsity:
> With your suggestions it's not obvious why model is not compressed with intermediate level of compression.
30% of sparsity is not compressed model, but with 60% compressed. It looks confusing.

Will we sacrifice it?

[the remark this answers is not in the document]
Comment (ljaljushkin) [reply examples/classification/main.py:225]: > With `is_compressed()` the logic of saving the best checkpoint can be simplified to the following one:
> 
> * We don't save or track best accuracy and checkpoint until `is_compressed` is true
> * When it gets true we start doing that
> 
> ```
> if compression_ctrl.is_compressed() and acc1 > best_acc1:
>     best_acc1 = acc1
>     save(...)
> ```

It's not that simple in reality by two reasons:

1. best_acc1 must be tracked from the very beginning for correct plateau schedule:
```
   if isinstance(lr_scheduler, ReduceLROnPlateau):
       lr_scheduler.step(best_miou)
```
2. best_acc1 is also needed for metrics, which are claimed to be helpful during the training along with compression level:
> You can run it as usual and monitor your original model's metrics and/or compression algorithm metrics and balance model metrics quality vs. level of compression.

We still need to save and track best_acc1 and catch the moment when model became compressed after uncompressed.
So there is NO difference with current approach, except for an issue:
> With your suggestions it's not obvious why model is not compressed with intermediate level of compression.
30% of sparsity is not compressed model, but with 60% compressed. It looks confusing.

**Comment by ljaljushkin** [comment]
> Should document this new functionality in Usage.md (or a more suitable location), so that users are made aware of it when trying to learn basic usage of NNCF.

Done
